algebird-core/src/main/scala/com/twitter/algebird/experimental/BloomFilter.scala

```diff
@@ -0,0 +1,413 @@
+    override val numBits: Int = numHashes
+
+    override def toBitSet: BitSet = BitSet(hashToArray(item))
+
+    override def ++(other: Hash): Hash =
+      other match {
+        case Empty => this
+        case Item(otherItem) =>
+          Instance {
+            BitSet
+              .newEmpty(0)
+              .mutableAdd(hashToArray(item))
+              .mutableAdd(hashToArray(otherItem))
+          }
+        case _ => other + item
+      }
+
+    override def +(other: A): Hash = this ++ Item(other)
+
+    override def checkAndAdd(other: A): (Hash, ApproximateBoolean) =
+      if (other == item) {
+        (this, ApproximateBoolean.exactTrue)
+      } else {
+        (this + other, ApproximateBoolean.exactFalse)
+      }
+
+    override def contains(x: A): ApproximateBoolean = ApproximateBoolean.exact(item == x)
+
+    override def maybeContains(x: A): Boolean = item == x
+
+    override def size: Approximate[Long] = Approximate.exact[Long](1)
+  }
+
+  /*
+   * Bloom filter with multiple values
+   */
+  case class Instance(bits: BitSet) extends Hash {
+
+    /**
+     * The number of bits set to true
+     */
+    override def numBits: Int = bits.size.toInt
+
+    override val toBitSet: BitSet = bits
+
+    override def ++(other: Hash): Hash =
+      other match {
+        case Empty               => this
+        case Item(item)          => Instance(bits | BitSet(hashToArray(item)))
+        case Instance(otherBits) => Instance(bits | otherBits)
+      }
+
+    override def +(item: A): Hash = this ++ Item(item)
+
+    override def checkAndAdd(other: A): (Hash, ApproximateBoolean) =
+      (this + other, contains(other))
+
+    override def maybeContains(item: A): Boolean = {
+      val il = hashToArray(item)
+      var idx = 0
+      var found = true
+      while (idx < il.length && found) {
+        val i = il(idx)
+        if (!bits(i)) {
+          found = false
+        }
+        idx += 1
+      }
+      found
+    }
+
+    // use an approximation width of 0.05
+    override def size: Approximate[Long] =
+      BloomFilter.sizeEstimate(numBits, numHashes, width, 0.05)
+  }
+
+  implicit val monoid: Monoid[Hash] with BoundedSemilattice[Hash] =
+    new Monoid[Hash] with BoundedSemilattice[Hash] {
+      override val zero: Hash = Empty
+
+      /**
+       * Assume the bloom filters are compatible (same width and same hashing functions).  This
+       * is the union of the 2 bloom filters.
+       */
+      override def plus(left: Hash, right: Hash): Hash = left ++ right
+
+      override def sum(t: TraversableOnce[Hash]): Hash =
+        if (t.iterator.isEmpty) empty
+        else {
+          var bfItem: Item = null
+          val iter = t.iterator
+          var counter = 0
+          var bs = BitSet.newEmpty(0)
+
+          while (iter.hasNext) {
+            iter.next() match {
+              case Empty => ()
+              case bi @ Item(item) =>
+                bfItem = bi
+                val array = hashToArray(item)
```

Comment:
if we had a variant of hashToArray that did not allocate the result array, we could reuse the array. Like hashIntoArray(item, ary). That would likely make a real difference here since we can allocate one array at the beginning.